Implemented Max-Rects packing

soimy opened the pull request:
Major improvement:

- Implemented Max-Rects packing, which is more efficient than multi-bin-pack.
- Add Decimal round option for json simplification and image pixel snapping.

Bug fixed:

- Font edge rendering artifacts on msdf output
- Glyph bounding-box and x/y offset issue.
- Alpha channel for monochrome sdf output.

soimy committed:
Fix Yoffset and padding issue.

SDF distanceRange should be taken into account for padding and baseline computation.
Also for accurate and platform dependent metric computation, we should use os2 table sTypoAscender instead of font.ascender.

soimy committed:
Update alpha value for sdf format output

According to BMFont format, channel 15 means use all rgba channel for glyph.

Also, the recent commit on bounding box computing fix present weird positioning on msdfgen output image data, so I revert the first command fix.

soimy committed:
Update alpha value for sdf format output

According to BMFont format, channel 15 means use all rgba channel for glyph.

Also, the recent commit on bounding box computing fix present weird positioning on msdfgen output image data, so I revert the first command fix.

soimy committed:
Add eslintrc

soimy committed:
Add new option ‘roundDecimal’

Typically set it to 0 for pixel snapping

soimy committed:
Implementation of Max-Rects Packing Algorithm

soimy committed:
Add missing new font for test-generate.js

soimy committed:
Merge remote-tracking branch 'origin/bin-pack-algorithm' into msdfgen1.4

# Conflicts:
#	index.js

soimy committed:
Match msdfgen 1.4 sdf output

soimy committed:
Fix for the firstCommand bBox commit

msdfgen xOffset should be computed from boundBox.left + pad
Also x/y offset decimal round have been added for accurate result. (With small bitmap font, decimal >=1 is recommended for accurate result)

arilotter commented on index.js:
```diff
@@ -72,9 +72,12 @@ function generateBMFont (fontPath, opt, callback) {
-      context.fillStyle = '#ffffff';
-      context.fillRect(0, 0, canvas.width, canvas.height);
-      // context.clearRect(0, 0, canvas.width, canvas.height);
+      if(fieldType == "msdf") {
```
Please use `===` for equality checks

arilotter commented on .eslintrc:
```diff
@@ -0,0 +1,16 @@
+{
```
I'd rather use `prettier` to explicitly format code than eslint to enforce a code style.

arilotter commented on index.js:
```diff
@@ -136,6 +137,7 @@ function generateBMFont (fontPath, opt, callback) {
       },
       kernings: kernings
     };
+    if(roundDecimal != null) RoundAllValue(fontData, roundDecimal);
```
Use strict equality checking `!==`

arilotter commented on index.js:
```diff
@@ -248,3 +250,37 @@ function generateImage (opt, callback) {
   });
 }
+
+function RoundAllValue (obj, decimal = 0) {
```
function names should be camelCased

arilotter commented on index.js:
```diff
@@ -248,3 +250,37 @@ function generateImage (opt, callback) {
 }
+
+function RoundAllValue (obj, decimal = 0) {
+  for (var key in obj) {
```
use `let` and `const` rather than var

arilotter commented on index.js:
```diff
@@ -248,3 +250,37 @@ function generateImage (opt, callback) {
+
+function RoundAllValue (obj, decimal = 0) {
+  for (var key in obj) {
+    if (obj.hasOwnProperty(key)) {
```
Use 
```js
Object.keys(obj).forEach(key => {
    ...
}
```
 instead of 
```js
for (var key in obj) {
    if (obj.hasOwnProperty(key){
        ...
    }
}
```

arilotter commented on index.js:
```diff
@@ -248,3 +250,37 @@ function generateImage (opt, callback) {
+  for (var key in obj) {
+    if (obj.hasOwnProperty(key)) {
+      if (obj[key] !== null) {
+        if (typeof(obj[key]) == "object") {
```
use `===`

arilotter commented on index.js:
```diff
@@ -248,3 +250,37 @@ function generateImage (opt, callback) {
+}
+
+function isNumeric (n) {
+  return !isNaN(parseFloat(n)) && isFinite(n);
```
I believe `global.isNaN` coerces its argument to a number anyways, so the parseFloat call is unnecessary. Additionally, `isFinite(NaN)` will return false, so I believe use of this function can just be replaced with `isFinite`.

soimy replied:
This is based on this solution [Validate decimal numbers in JavaScript](https://stackoverflow.com/questions/18082/validate-decimal-numbers-in-javascript-isnumeric).
I'm new to js, not sure whether all nodejs version use ES6. If so we could use Number.isFinite() instead?

arilotter commented on index.js:
```diff
@@ -248,3 +250,37 @@ function generateImage (opt, callback) {
+  if(!("" + num).includes("e")) {
+    return +(Math.round(num + "e+" + scale)  + "e-" + scale);
+  } else {
+    var arr = ("" + num).split("e");
```
Use `let` and `const`

arilotter commented on lib/maxrectpacker.js:
```diff
@@ -0,0 +1,35 @@
+"use strict";
```
I think this would serve better as a separate node module that we could introduce as a dependency.

soimy replied:
I will try to package it in npm module after finished the fine-tuning of these codes. This is my first node.js project, still a lot to learn ;)

arilotter commented on index.js:
```diff
@@ -248,3 +250,37 @@ function generateImage (opt, callback) {
+function RoundAllValue (obj, decimal = 0) {
+  for (var key in obj) {
+    if (obj.hasOwnProperty(key)) {
+      if (obj[key] !== null) {
```
Maybe refactor this to
```js
if(typeof(obj[key]) === "object" && obj[key] !== null) {
    ...
}
so we don't have to do the null check if it's a number?

ethernet8023 reviewed (changes requested):
If we're using the Dimbo font to test instead of Roboto, we should remove the Roboto ttf.

soimy committed:
Fix msdf font edge artifacts

Main reason of these artifacts is because msdfgen 1.4 invert both msdf/sdf output channel, so the bg color for msdf atlas should be black now.

soimy committed:
Revert "Add eslintrc"

This reverts commit 5adb655825a6aae73d1c451de3b9ebaf64554de0.

soimy committed:
Change testing font to ‘Din-condensed.ttf’

This font will show glyph feather much better, also visual represent of kerning and alignment is more obvious.

soimy committed:
Code review fix

soimy commented:
Code review fixed.
Also find the cause of msdf font edge artifacts, fixed. 🍺
<img width="739" alt="font-rendertest" src="https://user-images.githubusercontent.com/1550237/29201450-eecbfe8a-7e90-11e7-96ac-0e0a493c5e57.png">

neurolabusc commented:
@soimy thanks for this terrific PR. The fonts are packed much more efficiently, and the edge artifacts are gone. Fantastic work! I did have one minor issue with running this on my 10.11.6 MacOS computer. Specifically, the executable generated the error `Reason: Incompatible library version: msdfgen.osx requires version 21.0.0 or later, but libfreetype.6.dylib provides version 19.0.0`. This was not a problem for me, as it was easy to recompile msdfgen. However, to make it easier for others, I would recommend building this on an older machine. I am happy to share my executable. I am very impressed by your work.

soimy commented:
@neurolabusc msdfgen 1.4 osx binary in this repo is built on freetype v2.8. 
In your case, u just need to run `brew update && brew upgrade freetype` to update your freetype to the newest version.
But you are right, for more user-friendly approach, we need to either: 
- use private dynamic library with msdf binary, just like the windows version.
- use msdfgen as a nodejs native add-on ( built with node-gyp ) as mentioned #3

soimy committed:
Use private dylib for msdfgen osx binary

soimy commented:
@neurolabusc Can you help me test the private dynamic library version of the msdf binary?
Just made a commit in my repo soimy/msdf-bmfont-xml@f154f92dbe60ef730f7e683066856f2303bafad2

soimy committed:
Bin packing algorithm moved to npm module

soimy committed:
Update osx binary’s local dylib

neurolabusc commented:
@soimy your inclusion of the additional libraries works. Thanks again. Future versions of my MRIcroGL and Surfice will benefit from your work.

soimy commented:
@neurolabusc Happy to hear that! Keep an eye on my [dev branch](https://github.com/soimy/msdf-bmfont-xml/tree/develop) for more feature and fixes.

arilotter commented on package.json:
```diff
@@ -39,6 +39,8 @@
     "canvas": "^1.4.0",
     "map-limit": "0.0.1",
+    "maxrects-packer": "^1.0.2",
+    "js2xmlparser": "^3.0.0",
```
Do we need this dependency for anything?

soimy replied:
@arilotter Sorry about that, it was leaked from another branch (Export xml format)

soimy committed:
Clean package.json

soimy commented:
@arilotter Sorry about that, it was leaked from another branch (Export xml format)
